File: leetcode/C++/push-dominoes.cpp
class Solution {
public:
    string pushDominoes(string dominoes) {
        int idx, next;
        vector<int> visit(dominoes.length(),0), v(dominoes.length(),0);
        queue<int> q;
        for(int i=0; i<dominoes.length(); ++i){
            if(dominoes[i]=='L') {
                visit[i]=1; v[i]=-1; q.push(i); 
            }
            else if(dominoes[i]=='R') {
                v[i]=visit[i]=1; q.push(i); 
            }
        }
        while(!q.empty()){
            idx=q.front(); q.pop();
            next=idx+v[idx];
            if(v[idx]==0) continue;
            if(next<0 || next>=dominoes.length()) continue;
            if(visit[next]&&visit[idx]+1>visit[next]) continue;
            visit[next]=visit[idx]+1;
            v[next]+=v[idx];
            q.push(next);
        }
        for(int i=0; i<dominoes.length(); ++i){
            if(v[i]==-1) dominoes[i]='L';
            else if(v[i]==1) dominoes[i]='R';
            else dominoes[i]='.';
        }
        return dominoes;
    }
};

File: leetcode/Python/push-dominoes.py
class Solution:
    def pushDominoes(self, dominoes: str) -> str:
        visit, val, sol = [0]*len(dominoes), [0]*len(dominoes), []
        idx, next_idx = 0, 0
        dq = collections.deque()
        for i in range(len(dominoes)):
            if dominoes[i]=='L':
                visit[i], val[i] = 1, -1
                dq.append(i)
            elif dominoes[i]=='R':
                visit[i]=val[i]=1
                dq.append(i)
        while dq:
            idx = dq.popleft()
            next_idx = idx+val[idx]
            if val[idx]==0 or (next_idx<0 or next_idx>=len(dominoes)) or (visit[idx]+1>visit[next_idx]>0):
                continue
            visit[next_idx]=visit[idx]+1
            val[next_idx]+=val[idx]
            dq.append(next_idx)
        for i in val:
            if i==-1:
                sol.append('L')
            elif i==1:
                sol.append('R')
            else:
                sol.append('.')
        return ''.join(sol)